File: AbstractFactory/python-code/factories/RoomFactory.py
from __future__ import annotations
from abc import ABC

from rooms.HardRoom import HardRoom
from rooms.NormalRoom import NormalRoom


class RoomFactory(ABC):
    def create_normal_room(self) -> NormalRoom:
        raise NotImplementedError

    def create_hard_room(self) -> HardRoom:
        raise NotImplementedError

File: JavaSource/src/main/java/ru/akhcheck/patterns/abstractfactory/factories/RoomFactory.java
package ru.akhcheck.patterns.abstractfactory.factories;

import ru.akhcheck.patterns.abstractfactory.rooms.HardRoom;
import ru.akhcheck.patterns.abstractfactory.rooms.NormalRoom;

/**
 * Абстрактная фабрика: создаёт семейство комнат одного уровня.
 *
 * <p>Отличие от фабричного метода в слове «семейство». Фабричный метод создаёт
 * один объект, абстрактная фабрика гарантирует, что созданные объекты подходят
 * друг другу: комнаты подвала не перемешаются с комнатами пещер.
 */
public interface RoomFactory {
    NormalRoom createNormalRoom();

    HardRoom createHardRoom();
}
